### PR TITLE
Feature: optimise mint/burn costs

### PR DESCRIPTION
**Pull request by failfmi**
**Detailed description**:
* Optimise cost of mint/burn functions

**Which issue(s) this PR fixes**:
Fixes #none

**Special notes for your reviewer**:
* Update solidity version due to compilation problem

**Checklist**
- [ ] Documentation added
- [x] Tests updated

### CODE REVIEW

**Review by Daniel-K-Ivanov — COMMENTED** [review]
- I think that we should add another parameter to the `Mint` event -> `serviceFee`

**Review by Holymir — APPROVED** [review]

**Review by Daniel-K-Ivanov — COMMENTED** [review]
The Burn event emits `sender`, `amount` and `receiver`. I think that we should change the event to:
`event Burn(address indexed sender, uint256 amount, uint256 serviceFee, bytes indexed receiver`)

This way, we will have visibility, how much was spend to service fees and how much we should send on the Hedera side. Atm we will simply send the `amount` which is not correct since we won't exclude the service fee.

`amount` could be the "clean" amount to receive after service fee or it could be the initial amount that the user burned, does not matter, important thing is that we need to have the `serviceFee`emitted

File: contracts/Bridge.sol [hunk 1]
```diff
@@ -169,6 +169,8 @@ contract Bridge is Governance, Pausable {
 
     /// @notice Claims the accrued fees of a member
     function claim() public {
```
Comment (Daniel-K-Ivanov) [inline contracts/Bridge.sol:171]: Let's be cautious here and add `amount` parameter + require `amount <= claimableFees[msg.sender]`

Reasoning:
If for some reason the calculations on how much there is to claim in total (`totalClaimableFees`) and the `claimableFees[msg.sender]` result in discrepancies, we don't want to lock the entire balance of the member. If he has the option to specify amount, he would be able to claim a certain amount and leave only the small discrepancy lost in the contract.

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]

**Review by Daniel-K-Ivanov — APPROVED** [review]

File: test/bridge.js [hunk 2]
```diff
@@ -725,4 +711,138 @@ describe("Bridge", function () {
...
+            const totalCheckpointsAfterUpdateMember = await bridgeInstance.totalCheckpoints();
+            assert(totalCheckpointsAfterUpdateMember.eq(totalCheckpointsBeforeUpdateMember.add(1)));
+
+            const feesAccruedForPreviousCheckpoint = await bridgeInstance.checkpointServiceFeesAccrued(totalCheckpointsBeforeUpdateMember);
+            assert(feesAccruedForPreviousCheckpoint.eq(feesAccruedForCheckpointBeforeUpdateMember));
+
+            const feesAccruedForNewCheckpoint = await bridgeInstance.checkpointServiceFeesAccrued(totalCheckpointsAfterUpdateMember);
+            assert(feesAccruedForNewCheckpoint.eq(expectedLeftoversForNewCheckpoint));
+
+            const aliceTotalClaimableFees = await bridgeInstance.claimableFeesFor(aliceMember.address);
```
Comment (Daniel-K-Ivanov) [inline test/bridge.js:818]: I think that it would be beneficial to test that Alice can claim here fees even though she is not a member now

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]

**Review by Daniel-K-Ivanov — APPROVED** [review]

**Pull request merged**